cmd/snap-confine/ns-support.c

```diff
@@ -365,21 +408,34 @@ static int sc_inspect_and_maybe_discard_stale_ns(int mnt_fd,
 			die("cannot join preserved mount namespace");
 		}
 		// Check if the namespace needs to be discarded.
-		//
+		eventfd_t value = SC_DISCARD_NO;
+		const char *value_str = "no";
+
 		// TODO: enable this for core distributions. This is complex because on
 		// core the rootfs is mounted in initrd and is _not_ changed (no
 		// pivot_root) and the base snap is again mounted (2nd time) by
 		// systemd. This makes us end up in a situation where the outer base
 		// snap will never match the rootfs inside the mount namespace.
-		bool should_discard =
-		    inv->is_normal_mode ?
-		    should_discard_current_ns(base_snap_dev) : false;
-
-		// Send this back to the parent: 2 - discard, 1 - keep.
+		if (inv->is_normal_mode
+		    && should_discard_current_ns(base_snap_dev)) {
+			value = SC_DISCARD_SHOULD;
+			value_str = "should";
+
+			// The namespace is stale, let's check if we must discard it because
+			// base snap has changed. We must do so even though the "view" of the
+			// snap mount namespace becomes fractured because otherwise snaps might
+			// run against the base snap declared by an older revision of the app
+			// snap, For more information see LP: #1819875
```

Comment:
Perhaps reword as something along the lines of: "_The namespace is stale so also check if we must discard it due to the base snap changing. If the base snap changed, we must discard since even though currently running processes from this snap will continue to see the old base, we want new processes to use the new base. See LP: #1819875 for details._"